engine/consensus/matching/engine.go

```diff
@@ -149,1085 +105,209 @@ func New(
-	startTime := time.Now()
-	approvalSpan := e.tracer.StartSpan(approval.ID(), trace.CONMatchOnApproval)
-	defer func() {
-		e.metrics.OnApprovalProcessingDuration(time.Since(startTime))
-		approvalSpan.Finish()
-	}()
-
-	log := e.log.With().
-		Hex("approval_id", logging.Entity(approval)).
-		Hex("block_id", approval.Body.BlockID[:]).
-		Hex("result_id", approval.Body.ExecutionResultID[:]).
-		Logger()
-	log.Info().Msg("result approval received")
-
-	// Check that the message's origin (as established by the networking layer) is
-	// equal to the message's creator as reported by the message itself. Thereby,
-	// we rely on the networking layer for enforcing message integrity via the
-	// networking key.
-	if approval.Body.ApproverID != originID {
-		return engine.NewInvalidInputErrorf("invalid origin for approval: %x", originID)
-	}
-
-	// check if we already have the block the approval pertains to
-	head, err := e.state.AtBlockID(approval.Body.BlockID).Head()
-	if err != nil {
-		if !errors.Is(err, storage.ErrNotFound) {
-			return fmt.Errorf("failed to retrieve header for block %x: %w", approval.Body.BlockID, err)
-		}
-		// Don't error if the block is not known yet, because the checks in the
-		// else-branch below are called again when we try to match approvals to chunks.
-	} else {
-		// drop approval, if it is for block whose height is lower or equal to already sealed height
-		sealed, err := e.state.Sealed().Head()
-		if err != nil {
-			return fmt.Errorf("could not find sealed block: %w", err)
-		}
-		if sealed.Height >= head.Height {
-			log.Debug().Msg("discarding approval for already sealed and finalized block height")
-			return nil
-		}
-
-		// Check if the approver was a staked verifier at that block.
-		err = e.ensureStakedNodeWithRole(approval.Body.ApproverID, head, flow.RoleVerification)
-		if err != nil {
-			return fmt.Errorf("failed to process approval: %w", err)
+// consumeEvents consumes events that are ready to be processed.
+func (e *Engine) consumeEvents() {
+	// Context:
+	// We expect a lot more Approvals compared to blocks or receipts. However, the level of
+	// information only changes significantly with new blocks or new receipts.
+	// We used to kick off the sealing check after every approval and receipt. In cases where
+	// the sealing check takes a lot more time than processing the actual messages (which we
+	// assume for the current implementation), we incur a large overhead as we check a lot
+	// of conditions, which only change with new blocks or new receipts.
+	// TEMPORARY FIX: to avoid sealing checks to monopolize the engine and delay processing
+	// of receipts and approvals. Specifically, we schedule sealing checks every 2 seconds.
+	checkSealingTicker := make(chan struct{})
+	defer close(checkSealingTicker)
+	e.unit.LaunchPeriodically(func() {
+		checkSealingTicker <- struct{}{}
+	}, 2*time.Second, 120*time.Second)
+
+	for {
+		select {
+		case event := <-e.receiptSink:
+			e.consumeSingleEvent(event)
```

Comment:
We handle most cases here through `consumeSingleEvent`(except for the sealing check). As we don't pass any context to `consumeSingleEvent` has to re-check the types and recover the information we already have here. 

To me, this adds unnecessary complexity (personal stylistic preference). Hence, I would like to suggest:
* in each case, we know precisely, that type of event we are expecting in the queue 
* lets do an unchecked cast and feed it directly into the corresponding method in code. 

Furthermore, I would like to suggest to make the following changes to `Core`:
* export methods we use in the engine. This provides some guidance what methods should be used externally. 
* In my opinion, core should handle all the expected errors internally. Any error that is escalated to the engine can be taken as a fatal problem. 
   - for now, you could just move this into `Core` https://github.com/onflow/flow-go/blob/823315497f38e8d508cd4acf361ad1691fe077c0/engine/consensus/matching/engine.go#L250-L257
   - in case `consumeEvents` receives an error from `Core`, we assume that the node's internal state is corrupted. In that case, we `log.Fatal().Err(err)` and are done.